### DISCUSSION TITLE
Stopwatch as stimulus in html-keyboard-response

### DISCUSSION THREAD

**Discussion by einGlasRotwein** [jspsych/jsPsych#1155] [Q&A] [answered]
Hello everyone,

I'm totally new to jsPsych and I was wondering whether there is a way of displaying the elapsed response time during a trial. I'm trying to implement something like a Psychomotor Vigilance Task: As soon as a stop watch appears on the screen and begins to count up, participants have to press a button as fast as they can and then the stopwatch stops.

So what I need is a "live display" of 'rt' recorded by 'html-keyboard-response' while the trial is running - but I'm afraid this data only exists **after** the trial has been completed?

I tried using a timer function (because the code for a timer was linked [in this discussion](https://github.com/jspsych/jsPsych/discussions/1149)). Naively, I just copy-pasted the [code from Stack Overflow] linked in the discussion mentioned before and modified it in such a way that I can get the counter running in the console log. See the code on GitHub [here](https://github.com/einGlasRotwein/jspsych_stuff/tree/main/PVT).

But how do I manage to get it displayed as a stimulus?

(Eventually, the counter will need to start from zero and count up, not the other way around. But I'm lacking some serious insights how JavaScript works generally, so it's baby steps here.)

Can you point me towards any resources? Thank you in advance!
Juli

**Comment by chrisbrickhouse** [comment, marked as answer]
Hi @einGlasRotwein ! I've submitted a pull request (see einGlasRotwein/jspsych_stuff#1) that fixes the timer [along with comments](https://github.com/einGlasRotwein/jspsych_stuff/commit/d3277eba823dc2c09d545cc031856bc0cfa317a8#comments) to explain the problems.

The code you copied works well, but it was not displaying because the timer was printed to the developer's console, not the `#time` element. So in the timer function, you needed to add `display.innerHTML = minutes + ":" + seconds;`

The other problem is that you started the timer with `on_start` which means the timer begins *before* the actual text is loaded and displayed. So when you try to retrieve it the first time, nothing is found and you get an error. Instead, you should use `on_load` which waits until after the HTML stimulus has loaded.

Hopefully that helps, and feel free to ask any follow up questions you might have!

**Reply by einGlasRotwein** [reply]
Thank you so much, that is amazing! I did not expect that much support, but I already noticed this forum is amazing.

It was just printed to the console because this was the only place where I was able to display it :-D
That clarified a lot, I will try and expand this further :-)

**Comment by einGlasRotwein** [comment]
A follow-up on this: I was now trying to have several stopwatch trials in a row, but I don't know how to manage the correct updating of the timer. I have an example with three trials (fixation, stopwatch, fixation, stopwatch ...) [here on GitHub](https://github.com/einGlasRotwein/jspsych_stuff/tree/main/PVT). I thought when I specify five minutes as the duration argument of the stopwatch, that is enough to reset the stopwatch every time.

Apparently, the stopwatch **does** reset and counts back as it's supposed to do. However, this only works properly for the first trial. From the second trial onwards, the initial stopwatch (i.e. the one from the first trial) is "flickering" through while the "current" stopwatch is counting back correctly. So, there seem to be some "leftovers" of the first stopwatch (and only the first) which was created. How do I manage to only show the current stopwatch in each trial?

Thank you in advance!

**Reply by becky-gilbert** [reply]
Hi @einGlasRotwein I didn't go through your code too closely, but it looks like you're using `setInterval` but not stopping it. This means that, after the first `setInterval` is called, the timer code will continue running during subsequent trials (which also start their own `setInterval` timers, so you'll have multiple timers running at once). What you can do is save your `setInterval` as a variable so that you can clear it at the end of the trial. See the [documentation and examples for clearInterval](https://www.w3schools.com/jsref/met_win_clearinterval.asp). I think the best place to stop the timer is in the `on_finish` function for each trial.

**Reply by einGlasRotwein** [reply]
Thank you so much for your answer. I was able to set the interval in a variable (i.e. my experiment worked like before, with multiple stopwatches running). However, adding `clearInterval` to `on_finish` did not work. The first trial runs just fine, but after that, I get the error 

```
Uncaught ReferenceError: timer_interval is not defined
    at Object.on_finish (PVT.html:80)
```

Is that because I define the variable within a function, so I can't access it at `on_finish`? Defining `timer_interval` outside of the `StartTimer` function did not work for me (the counter then immediately jumps to 0:00).

The  current version of the file ist PVT.html [here](https://github.com/einGlasRotwein/jspsych_stuff/tree/main/PVT).

Sorry for the noob questions - can you recommend tutorials that deal with the logic behind displaying/updating/resetting html-elements?

**Reply by einGlasRotwein** [reply]
Additionally, I tried to simplify the function I had copied from Stack Overflow so I could specify the `timer_interval` outside of a function. In that case however, the timer immediately jumps to 0:00. This happens in the file PVT2.html [here](https://github.com/einGlasRotwein/jspsych_stuff/tree/main/PVT).

**Reply by becky-gilbert** [reply]
> Is that because I define the variable within a function, so I can't access it at on_finish?

Ah, sorry, yes that's exactly right. 

I think defining the `timer_interval` variable outside of the function should work. So you had the right idea with your second attempt - the reason it didn't work is because you not only defined the `timer_interval` variable (line 14) but also started the `setInterval` function (line 17) in your main script. 

Instead, I used the code from your first attempt but just moved the timer definition to the global scope. That way the `timer_interval` variable can be set to be the `setInterval` timer in one function and cleared with `clearInterval` from within a different function. Does this work for you?
```javascript
   var timer_interval; // declare the variable in global scope, but leave it undefined so that you don't start the timer yet
    // Timer function copied from:
    // https://stackoverflow.com/questions/20618355/the-simplest-possible-javascript-countdown-timer
    function startTimer(duration) {
        var start = Date.now(),
            diff,
            minutes,
            seconds;

        // Interval in a variable so it can be reset
        timer_interval = setInterval(timer, 1000); // now set the value of the timer_interval variable, which also starts the timer

        function timer() {
            // get the number of seconds that have elapsed since 
            // startTimer() was called
            diff = duration - (((Date.now() - start) / 1000) | 0);

            // does the same job as parseInt truncates the float
            minutes = (diff / 60) | 0;
            seconds = (diff % 60) | 0;

            minutes = minutes < 10 ? "0" + minutes : minutes;
            seconds = seconds < 10 ? "0" + seconds : seconds;
            display = document.querySelector('#time');
            display.innerHTML = minutes + ":" + seconds;

            if (diff <= 0) {
                // add one second so that the count down starts at the full duration
                // example 05:00 not 04:59
                start = Date.now() + 1000;
            }
        };
        // we don't want to wait a full second before the timer starts
        timer();
        // setInterval(timer, 1000);
    }

    var timeline = [];
    var n_trials = 3;

    var start_screen = {
        type: 'html-keyboard-response',
        stimulus: 'So it begins.<p>' +
            'Press a button to continue.<p>' +
            'Press a button every time the clock appears.',
        data: { condition: 'start_screen' }
    }
    timeline.push(start_screen)

    for (var i = 0; i < n_trials; i++) {
        var fixation = {
            type: 'html-keyboard-response',
            stimulus: '<span fontsize>+</span>',
            choices: jsPsych.NO_KEYS,
            trial_duration: 800,
            data: { condition: 'fixation' }
        }
        timeline.push(fixation)

        var stopwatch = {
            on_load: function () {
                var fiveMinutes = 60 * 5
                startTimer(fiveMinutes);
            },
            type: 'html-keyboard-response',
            stimulus: '<div><span id="time">05:00</span></div>',
            data: { condition: 'stopwatch' },
            on_finish: function () {
                clearInterval(timer_interval); // this should now work without producing an error because timer_interval was declared outside of a function
            }
        }
        timeline.push(stopwatch)
    }

    jsPsych.init({
        timeline: timeline,
        on_finish: function () {
            jsPsych.data.displayData();
        }
    })
```

**Reply by einGlasRotwein** [reply]
Thank you so much, this works perfectly! I'm starting to understand the whole JavaScript logic a bit better.
